File: frontend/src/app/services/sensor.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { environment } from '../../environments/environment';
import { MOCK_MODE } from '../config/mock.config';

export interface LecturaDto {
  tiempo: string;
  aceleracion: number;
  velocidad: number;
  desplazamiento: number;
  canal: string;
  componente: string;
}

export interface SensorDataDto {
  idSensor: number;
  nombre: string;
  marca: string;
  modelo: string;
  tipoSensor: string;
  descripcion: string;
  lecturas: LecturaDto[];
}

@Injectable({
  providedIn: 'root'
})
export class SensorService {
  private apiUrl = `${environment.apiUrl}/sensores`;

  constructor(private http: HttpClient) { }

  getSensorsByEstacion(estacionId: number): Observable<SensorDataDto[]> {
    if (MOCK_MODE) {
      return of(this.getMockSensors(estacionId, 6));
    }
    return this.http.get<SensorDataDto[]>(`${this.apiUrl}/estacion/${estacionId}`);
  }

  getSensorsWithRecentReadings(estacionId: number, horas: number = 24): Observable<SensorDataDto[]> {
    if (MOCK_MODE) {
      return of(this.getMockSensors(estacionId, horas));
    }
    return this.http.get<SensorDataDto[]>(`${this.apiUrl}/estacion/${estacionId}/lecturas-recientes?horas=${horas}`);
  }

  private getMockSensors(estacionId: number, horas: number): SensorDataDto[] {
    const now = new Date();
    const makeLecturas = (canal: string, componente: string) => {
      const puntos = Math.min(horas * 12, 120);
      const lecturas = [] as LecturaDto[];
      for (let i = puntos; i >= 0; i--) {
        const t = new Date(now.getTime() - i * 5 * 60 * 1000).toISOString();
        const base = 0.010 + Math.random() * 0.008;
        lecturas.push({
          tiempo: t,
          aceleracion: parseFloat((base).toFixed(6)),
          velocidad: parseFloat((base * 0.25).toFixed(6)),
          desplazamiento: parseFloat((base * 0.08).toFixed(6)),
          canal,
          componente
        });
      }
      return lecturas;
    };

    return [
      {
        idSensor: estacionId * 10 + 1,
        nombre: 'Acelerómetro Triaxial',
        marca: 'Kinemetrics',
        modelo: 'Episensor ES-T',
        tipoSensor: 'Acelerógrafo',
        descripcion: 'Sensor principal de la estación',
        lecturas: [
          ...makeLecturas('Canal Z', 'Z'),
          ...makeLecturas('Canal N', 'N'),
          ...makeLecturas('Canal E', 'E')
        ]
      }
    ];
  }
}

File: backend/src/main/java/com/seismic/service/SensorService.java
package com.seismic.service;

import com.seismic.dto.SensorDataDto;
import com.seismic.entity.Estacion;
import com.seismic.entity.Lectura;
import com.seismic.entity.Sensor;
import com.seismic.repository.EstacionRepository;
import com.seismic.repository.SensorRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class SensorService {

    private final SensorRepository sensorRepository;
    private final EstacionRepository estacionRepository;

    public List<SensorDataDto> getSensorsByEstacion(Integer estacionId) {
        Estacion estacion = estacionRepository.findById(estacionId)
                .orElseThrow(() -> new RuntimeException("Estación no encontrada"));

        List<Sensor> sensores = sensorRepository.findByEstacion(estacion);
        
        return sensores.stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    public List<SensorDataDto> getSensorsWithRecentReadings(Integer estacionId, int horasAtras) {
        LocalDateTime desde = LocalDateTime.now().minusHours(horasAtras);
        
        List<Sensor> sensores = sensorRepository.findByEstacionIdWithRecentReadings(estacionId, desde);
        
        return sensores.stream()
                .map(sensor -> convertToDtoWithReadings(sensor, desde))
                .collect(Collectors.toList());
    }

    private SensorDataDto convertToDto(Sensor sensor) {
        String nombreSensor = generateSensorName(sensor);
        
        return SensorDataDto.builder()
                .idSensor(sensor.getIdSensor())
                .nombre(nombreSensor)
                .marca(sensor.getMarca())
                .modelo(sensor.getModelo())
                .tipoSensor(sensor.getTipoSensor().getNombre())
                .descripcion(sensor.getDescripcion())
                .build();
    }

    private SensorDataDto convertToDtoWithReadings(Sensor sensor, LocalDateTime desde) {
        String nombreSensor = generateSensorName(sensor);
        
        List<SensorDataDto.LecturaDto> lecturas = sensor.getCanales().stream()
                .flatMap(canal -> canal.getLecturas().stream()
                        .filter(lectura -> lectura.getTiempo().isAfter(desde))
                        .map(lectura -> SensorDataDto.LecturaDto.builder()
                                .tiempo(lectura.getTiempo())
                                .aceleracion(lectura.getAceleracion())
                                .velocidad(lectura.getVelocidad())
                                .desplazamiento(lectura.getDesplazamiento())
                                .canal(canal.getTipoCanal().getCodigo())
                                .componente(canal.getComponente().getNombre())
                                .build()))
                .collect(Collectors.toList());

        return SensorDataDto.builder()
                .idSensor(sensor.getIdSensor())
                .nombre(nombreSensor)
                .marca(sensor.getMarca())
                .modelo(sensor.getModelo())
                .tipoSensor(sensor.getTipoSensor().getNombre())
                .descripcion(sensor.getDescripcion())
                .lecturas(lecturas)
                .build();
    }

    private String generateSensorName(Sensor sensor) {
        StringBuilder nombre = new StringBuilder();
        
        // Usar tipo de sensor como base
        nombre.append(sensor.getTipoSensor().getNombre());
        
        // Agregar marca si está disponible
        if (sensor.getMarca() != null && !sensor.getMarca().isEmpty()) {
            nombre.append(" ").append(sensor.getMarca());
        }
        
        // Agregar modelo si está disponible
        if (sensor.getModelo() != null && !sensor.getModelo().isEmpty()) {
            nombre.append(" ").append(sensor.getModelo());
        }
        
        // Si no hay información suficiente, usar ID
        if (nombre.toString().equals(sensor.getTipoSensor().getNombre())) {
            nombre.append(" #").append(sensor.getIdSensor());
        }
        
        return nombre.toString();
    }
}
